homeassistant/components/directv/media_player.py

```diff
@@ -66,7 +66,7 @@
 async def async_setup_entry(
     hass: HomeAssistant,
     entry: ConfigEntry,
-    async_add_entities: Callable[[list, bool], None],
+    async_add_entities: AddEntitiesCallback,
```

Comment:
I am not sure this is a real improvement, because you hide the call parameters.

Reply:
Improvement here is you don't need to write complex `Callable[[Iterable[Entitity], bool], None]` type and can use imported type which already has everything.

Reply:
Yes but hidden ! If I want to make it simple, I simply do not add the : <type> which a number of integrations do.

Reply:
Do not add what? How is this different from `ConfigType` for example? You may say it hides the fact it is a dict.

Reply:
add the type, you can find integrations that just write "async_add_entities," without any type at all, see f.x. amp/air_quality.py

Reply:
Python is not a strongly typed language, it is up to the developer to use (or not use) a level of type checking.

Reply:
Well, this PR doesn't aim to add type annotation for this callback everywhere. I replaced only existing annotations.

Reply:
That is the point, I am pretty sure we could do the same with a high number of parameters used, and in the end we would not have a better or more stable product, because there would still be many instances without type checking at all. If we were to enforce type checking, it would lead to a slightly more stable product, but that is an architectural decision.

Reply:
If you really want to help to make it more stable and robust, please help to fix something from list of ignored components. See discussion in https://github.com/home-assistant/core/pull/49270

This is intended to simplify enabling type checks for more components, not to make it harder.

Reply:
Btw, `modbus` which you contributed to is listed there:
https://github.com/home-assistant/core/blob/e85d58c3a1bf9fc977cf7af5bee1fd0bb3477971/script/hassfest/mypy_config.py#L138
I'd be glad if you fix type annotations for it.

Reply:
That list contains nearly all (if not all) integrations, and fixing modbus with your type would not remove modbus from the list.

We have no decision to enforce type checking, so that list can at best be seen as a recommendation. Enforcing type checking would be a major undertaking which would cause changes to a very big % of the code base, and I for one do not see the need. Being a very long time C++ programmer I see some beauty in leaving type checking up to the single developer.

Reply:
That list contains 243 out of 987 integrations.

I don't mean fixing it with my type. I mean fixing existing errors it has. To help to make integration and project more stable as you mentioned. Because it has some type hints which are not valid which means potential bugs.

That list doesn't actually enforce typing, it excludes modules which have broken type hints. If you care about better and stable project, I see the value in fixing this.

Covering such a large project with type checks is collective effort, no single person can do that. For sure it can't be enabled in one go, that will be slow and gradual process.

Reply:
But before covering this project with type checks, it would be nice to have it as a decision, in the architecture repo :-)

Of course the list does not enforce type checking, but "broken type hints" is a very strong wording e.g. using Callable is not broken nor invalid or an error, it is just not complete. I do not think the vast majority in that list have invalid type hints but just incomplete type hints, which are quite ok.

Reply:
We should enforce rules where it really matters. Enforcing that using "Callable" is not ok, while not doing type checking at all is ok does not seem very logical.

Reply:
Currently typing is encouraged but not required: https://developers.home-assistant.io/docs/development_guidelines/#typing

Just using `Callable` is not an error, it's another issue. But "broken type hints" means mypy returns errors when basic (not strict) type checking is enabled for them. All components in that list have some kind of existing errors.

Reply:
This is list of errors for `modbus`:
```
homeassistant/components/modbus/binary_sensor.py:87: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[str, Any]")  [assignment]
homeassistant/components/modbus/binary_sensor.py:94: error: Name 'hub' already defined on line 92  [no-redef]
homeassistant/components/modbus/switch.py:121: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[str, Any]")  [assignment]
homeassistant/components/modbus/switch.py:128: error: Name 'hub' already defined on line 126  [no-redef]
homeassistant/components/modbus/switch.py:132: error: Argument 1 to "append" of "list" has incompatible type "ModbusRegisterSwitch"; expected "ModbusCoilSwitch"  [arg-type]
homeassistant/components/modbus/switch.py:302: error: Return value expected  [return-value]
homeassistant/components/modbus/sensor.py:123: error: Incompatible types in assignment (expression has type "None", variable has type "Dict[str, Any]")  [assignment]
homeassistant/components/modbus/sensor.py:180: error: Name 'hub' already defined on line 178  [no-redef]
homeassistant/components/modbus/climate.py:245: error: Return value expected  [return-value]
homeassistant/components/modbus/climate.py:258: error: Incompatible types in assignment (expression has type "float", variable has type "Tuple[Any, ...]")  [assignment]
homeassistant/components/modbus/climate.py:262: error: Incompatible types in assignment (expression has type "float", variable has type "str")  [assignment]
homeassistant/components/modbus/climate.py:265: error: Incompatible return value type (got "str", expected "Optional[float]")  [return-value]
Found 12 errors in 4 files (checked 3899 source files)
```

Reply:
This discussion is far beyond the scope of this PR, let us stop here.

Reply:
Sorry I did not see you "error" list. It arrived after my last comment, but since you claim it is "errors" I need to correct your view.

f.x. sensor.py line 178-180:
```
if CONF_HUB in entry:
    # from old config!
    hub: ModbusHub = hass.data[MODBUS_DOMAIN][entry[CONF_HUB]]
else:
     hub: ModbusHub = hass.data[MODBUS_DOMAIN][discovery_info[CONF_NAME]]
```
is referred as "error: Name 'hub' already defined on line 178  [no-redef]" which is not correct, hub is only defined once but in 2 different flavours.

Line 123
```
    config = None
```
is referred to as an error because "config" is an argument "config: ConfigType", the argument may not be None so the argument declaration, however setting it to None is on purpose and not an error. Remember if I had declared the argument "config," it would not have been flagged. 

I know mypy detects a number real cases, but it also flags a lot of false positives.

Please be a bit careful to claim code to have errors, without actually looking at the code, the error might be in mypy !

Reply:
Typing errors doesn't need to be real errors. That code does have an typing issue.
Mypy don't want types defined twice. The typ of hub should be defined before the two selectable assignments. Not a mypy bug. 

It would have allowed you casting the type before assignment rather than defining hub's type twice.

Reply:
But that code is valid, it is not defined twice, and remember if I remove the type mypy will not flag it, but would you call the code more readable or more stable (without the type).

Reply:
Like I said, you could have casted it: cast(ModbusHub, xxx)
Or you could have declared it.

```
hub: MudbusHub
if a: hub = xxx
else: hub = yyy
```

It's just defining the type twice that is invalid. It's not invalid python, it's invalid type information.

You can't declare the type of a variable twice in many other types languages either.

Reply:
I know I could, but I could also have left out the type declaration, which are easier and just as correct, but a little less safe. 

Anyhow this is way off-topic for this PR.